### DISCUSSION TITLE
Feedback on Relm4's view macro.

### DISCUSSION THREAD

**Discussion by edfloreshz** [Relm4/Relm4#461] [Ideas]
Please read #450 for context.

**Comment by AaronErhardt** [comment]
Let's summarize the previous discussions about [declarative](https://github.com/ejaa3/declarative) and what features could be implemented in Relm4. 

## Architectural changes

Declarative does not require a trait nor a built-in runtime for interactivity, whereas Relm4 requires you to use messages to modify state. The underlying code generation is not fundamentally different, but rather it is about where the code is put by the macro. Declarative (ab)uses Rust's attribute macros to do something like this:

```rust
#[view { /* enter your widgets here */ }]
fn some_function() {
    // This is still part of the macro
}
```

This solves the problem that you need both the `view` macro and a location to paste the generated code by using the attributes of the macro for the view macro and the actual body of the macro for a function body. This gives the macro the power to control two separate inputs at once. While this is quite clever, it is surely not very idiomatic and confusing for beginners because if you put the function not immediately after the macro, the code will break.

Instead, I would rather like to explore using different traits, maybe a even more reduced version of the old [`MicroComponent`](https://aaronerhardt.github.io/docs/relm4/relm4/struct.MicroComponent.html) approach. This trait should allow users to write their own abstractions around it.

## Syntax changes

If you use a `!` after a widget declaration, declarative will use the builder pattern for setting values until you use a `#` which most notably removes the "set_" prefix in a lot of methods. This was already possible in Relm4, but only with the actual Rust syntax and not in declarative style. I like the idea, but I'm not quite happy with the syntax yet. `!` and `#` are not the most visible markers and they don't make sense for someone without prior knowledge of the macro.

Attributes are replaced by custom syntax or are instead written as lifetimes. The reduction of attributes makes the code look more elegant, but does not necessarily improve the readability. This alone should probably have it's own dedicated discussion and I don't have time to summarize everything, currently.

In declarative, arguments are passed by specifying the method after the widget type and then using `#` to mark the location of the widget inside of the method call. I like the idea, but I'm not sure if the position after the widget type is better.

And also `signal => closure` was changed to `@signal = closure` and uses a `clone` macro instead of brackets `[]` for cloning values. I don't think this is better, but this is mostly about personal preference and not a big feature to implement either.

## Conditional widgets++

Conditional widgets (using `if` or `match` in the view macro) are really handy, but only support dynamic use cases. If you want to only create a certain widget at initialization, you have to do it manually. In the future, conditional widgets should support both dynamic and static use-cases, they could even have the same syntax with only an attribute to differentiate both variants. 

## Replacing syn

This has nothing to do with the previous discussions, but I'd like to replace syn in the future. Syn still feels like a workaround even after years of development and has a lot of rough edges. To be fair, this has also to do with the limited APIs of the Rust compiler, but I think it's about time to challenge the status quo. 

I think a successor to syn should be more opinionated. Good macros should always have three stages: parsing, internal logic and code generation. Syn makes it possible to mix everything together, but parsing and code generation could be largely automated by macros and only the internal logic should be imperative code.

Syn supports parsing files, but doesn't parse comments and blank lines which makes it basically impossible to build a formater for Relm4 without duplicating the parsing logic of the macro. Together with the automated parsing logic, it should be possible to define where you want line-breaks and offer formatting at almost no extra effort for the user.

# Conclusion

There is still a lot more to be discussed, but I don't have more time now, so feel free to complete this summary and share your thoughts! We might even need to split this discussion into smaller topics to keep track of everything in the future.

**Reply by ejaa3** [reply]
## View syntax

The `#[view]` attribute is just an alternate syntax for the `block!` macro. Its advantages are to separate view from logic and decrease indentation. The full syntax is:

~~~ rust
use declarative::view;

#[view { /* widgets here */ }] //  first view
#[view { /* widgets here */ }] // second view

fn function() { // or an impl
    expand_view_here!(); // first view
    let closure = || expand_view_here!() // second view
}
~~~

Although normally a single view would be used.

> If you put the function not immediately after the macro, the code will break.

I assume you figured this out by using `#[relm4::component]` after the `impl`. (?

Logically the `#[view]` attribute can also be inside an `impl`, as well as it must be followed by a function or something else because that is how attribute macros work. If what you want is to put the view in the function, use `block!`:

~~~ rust
use declarative::block as view;

fn function() {
    view! { /* widgets here */ } // first view
    
    let closure = || {
        // `block!` (ie `view!`) is not autowrapped
        // in braces like `expand_view_here!`
        view! { /* widgets here */ } // second view
    };
}
~~~

## Builder mode

> If you use a ! after a widget declaration, declarative will use the builder pattern

Actually the `!` must go before the brace because after the declaration of a widget there can be `mut`, its name or an #interpolation. And the marker is not just `#`, but four: `#:`, `#;`, `@:` and `@;`. I try to explain the differences in the [first example](https://github.com/ejaa3/declarative/blob/main/examples/a_basics.rs).

For now I am not interested in making the marks the most visible (precisely I want to avoid the boilerplate) and that they make sense to those who do not know the macro (the examples are documented for a reason). I use `!` because it looks like another macro is being invoked (actually yes, the `builder_mode!` macro), and `#` and `@` are used because the same symbols are used for other things. Rust should report a misuse.

Also builder mode not only avoids `set_` (gtk-rs thing), but also wrapping in `Some` (gtk-rs thing) and compositing is possible (#interpolating with builder methods). For example, in Relm4 one would do:

~~~ rust
relm4::view! {
    my_flap = adw::Flap {
        set_flap_position: gtk::PackType::End
        set_fold_policy: adw::FlapFoldPolicy::Never
        
        #[wrap(Some)]
        #[name(the_flap)]
        set_flap = &gtk::Box {
            // content
        }
        #[wrap(Some)]
        #[name(the_content)]
        set_content = &gtk::Box {
            // content
        }
    }
}
~~~

In `declarative` without builder mode (there are `set_` and `Some`):

~~~ rust
declarative::block! {
    adw::Flap my_flap {
        set_flap_position: gtk::PackType::End
        set_fold_policy: adw::FlapFoldPolicy::Never
        
        gtk::Box the_flap #set_flap(Some(&#)) {
            // content
        }
        gtk::Box the_content #set_content(Some(&#)) {
            // content
        }
    }
}
~~~

In `declarative` with builder mode:

~~~ rust
declarative::block! {
    adw::Flap my_flap !{
        flap_position: gtk::PackType::End
        fold_policy: adw::FlapFoldPolicy::Never
        
        gtk::Box the_flap #flap(&#) {
            // content
        }
        gtk::Box the_content #content(&#) {
            // content
        }
    }
} // in this case there was no need to use #: #; @: or @;
~~~

## Attributes

> The reduction of attributes makes the code look more elegant, but does not necessarily improve the readability. This alone should probably have it's own dedicated discussion and I don't have time to summarize everything, currently.

Don't spend time on that either. While for Relm4 the attributes are just custom syntax, for `declarative` the attributes are the real Rust ones. So, I can do something like:

~~~ rust
declarative::block! {
    adw::ApplicationWindow {
        #[cfg(debug_assertions)] // real attribute
        add_css_class: "devel"
        
        gtk::Box #set_content(Some(&#)) {
            set_orientation: gtk::Orientation::Vertical
            
            #[cfg(feature = "with-sidebar")] // real attribute
            adw::Flap sidebar #append(&#) {
                // content
            }
        }
    }
}
~~~

You can see that I only add the [`devel`] class when debugging the app, and that the `adw::Flap` and anything inside it (properties, children, bindings, etc.) will not be compiled unless I enable the feature `with-sidebar`.

[`devel`]: https://gnome.pages.gitlab.gnome.org/libadwaita/doc/1-latest/style-classes.html#development-window

I try to support attributes anywhere, which you can try by expanding the example [z_besides] (doc comments are attributes). If it happens to you that the attribute macro doesn't fully expand, you can try moving the content of the attribute to `block!` and expanding `block!`.

[z_besides]: https://github.com/ejaa3/declarative/blob/main/examples/z_besides.rs

## Composition and returns

> In declarative, arguments are passed by specifying the method after the widget type and then using # to mark the location of the widget inside of the method call. I like the idea, but I'm not sure if the position after the widget type is better.

Before everything `declarative` has no syntactic sugar for adding children (a trait like `SetChildExt`; it had it but I removed it), and it's nice to see the type of the child at the beginning.

In addition, `declarative` calls methods like `method: "args"`, assigns struct fields like `field = "value"`, edits a struct field like `field => { /* edit */ }`, and calls a functional field like `functional := "args"`, so it seems that only the `->` arrow is left free to add children. So the syntax would be: `attach[&#, 1, 2, 3, 4] -> gtk::Widget { }` all the time. Still I think the arrow could be used for something else in the future although really only the brackets are enough.

However we have not taken into account the methods that return. To start the interpolation doesn't have to go after the widget declaration, but also inside the scope like so:

~~~ rust
gtk::Stack {
    gtk::Box {
        set_orientation: gtk::Orientation::Vertical
        set_spacing: 6
        #add_child(&#) // any position within the scope
        
        adw::Bin #attach(&#) { }
    }
}
~~~

In `declarative` I edit the return of calls to methods, or functional fields, or `#interpolations(#)` or `@extensions(#)` with the same keyword (`'back`). So I edit the above like so:

~~~ rust
gtk::Stack {
    gtk::Box {
        set_orientation: gtk::Orientation::Vertical
        set_spacing: 6
        
        #add_child(&#) 'back returned_page {
            set_name: "page_name"
            set_title: "Page title" }
        }
        
        adw::Bin #attach(&#) { }
    }
}
~~~

Currently in Relm4 I am forced to edit the return always at the end. [More about `'back` here](https://github.com/ejaa3/declarative/blob/main/examples/i_returns.rs).

## Signals and cloning

> And also signal => closure was changed to @signal = closure

In fact what is `signal => closure` in Relm4, is just the usual `method: arguments` in `declarative`. The `@variable_name = expression_with_bindings` syntax is for consuming bindings.

> and uses a clone macro instead of brackets [] for cloning values. I don't think this is better, but this is mostly about personal preference and not a big feature to implement either

For now I am not sure about cloning with brackets from the view macro so I dedicate a separate `clone!` macro. I can conveniently clone per method argument (Relm4 only allows one argument if I use `signal[clones] => closure`, which seems sufficient for gtk-rs). However, gtk-rs has its own `clone!` macro with dedicated features like weak clone.

If you're curious, the [struct of a component] in `declarative` when using channels (the Elm pattern) basically consists of the main widget plus its sender, and if you only need to clone the sender field, just use `component.sender` with `clone!` according to [the documentation].

[struct of a component]: https://github.com/ejaa3/declarative/blob/main/examples/g_components.rs#LL17C1-L20C2

[the documentation]: https://docs.rs/declarative/0.3.2/declarative/macro.clone.html

## Conditional widgets

`declarative` already supports `if` and `match` conditionals without reactivity (you refer to this as “static case”). Not only multiple widgets per condition are possible, but also multiple field assignments or calls to methods or functional fields. I haven't illustrated this properly, but you can see it in the same [z_besides] example, at the end.

Relm4's “reactive conditional” (or “dynamic case”) is really syntactic sugar for a `gtk::Stack` with automatic `set_visible_child`, which allows you to destructure states (which `#[watch]` and `# [track]` cannot do I think). However, the `'bind` of `declarative` allows conditionals and destructing on it, and `bindings!()` allows expansion anywhere, so I can explicitly simulate Relm4's “reactive conditionals”. Since I'm not sure about adding that syntactic sugar to `declarative`, I made a Rust project where I simulate it in two ways:

[comparison.zip](https://github.com/Relm4/Relm4/files/11561483/comparison.zip)

How strange that Rust throws a warning because I use `skip_init` in Relm4.

The project consists of examples, each suffixed with `decl` for `declarative` or `relm4` for Relm4. It has a counter example that I only did as an example to model the ones that come.

Now the interesting example is `conditional_relm4` where I show the use of match, but it could also have been `if`. Then there is the `conditional_decl` example where I simulate Relm4's `match` with declarative, but I am evaluating three conditions (two `if` and one `match`). The last example is `conditional_decl_alt`, where I give an alternative where I only use a single `match` condition.

**Comment by ejaa3** [comment]
Greetings. You will excuse the idiomatic errors since I use the translator a lot.

I guess I'm correct in saying that while Relm4 focuses on giving more through more features, the `declarative` approach is to give the most with as few features as possible, sort of the KISS principle. However `declarative` is reduced to simple macros to create complex and reactive views in a simple way, and even so, while components, templates and the Elm architecture are indispensable features for Relm4 (so code is dedicated for it in the library), for `declarative` are just usage patterns of the macro. That is why the `declarative` [examples] do not focus solely on showing what the macro does, but rather on usage patterns.

[examples]: https://github.com/ejaa3/declarative/tree/main/examples

So `declarative` doesn't seem like a full competition to Relm4. In fact it is possible that in the future Relm4 will drop some of its macros in favor of `declarative` ones, which means that Relm4 will use fewer traits and make its structures wrappers easy to use in a pure gtk-rs app, perhaps a type-safe API for settings and actions like the old API which is easier to maintain than the new one, or factories. They should automatically integrate well with `declarative` views due to their `#interpolations(#)` and `@extensions(#)`.

Eventually I will answer some issues with `declarative` in this space, as well as show my issues with Relm4.

**Comment by AaronErhardt** [comment]
Thanks @ejaa3 for your comments! I wrote down some of my thoughts here, but I had limited time, so some ideas might not be complete yet. 

### View syntax

I don't think Relm4 should use attribute macros like declarative does it. When looking at the [rust lang reference](https://doc.rust-lang.org/reference/procedural-macros.html#attribute-macros), it becomes clear that attribute macros are supposed to work similar to built-in attributes and should have only few additional parameters (also, declarative even breaks syntax highlighting in some editors because nobody even considered the possibility to (ab)use macros like this). I don't know a single crate that does things differently except for the current version of declarative. Therefore, I think we should always wrap the macro around a trait implementation to avoid unidiomatic code, roughly something like this:

```rust
#[some_view_macro(some_params)]
impl SomeTrait for Data {
    view! {
        // widgets...
    }

    fn some_method() {
        expand_view_here!();
    }
}
```

### Placing names

I think allowing to name widgets after the type like declarative and [blueprint](https://jwestman.pages.gitlab.gnome.org/blueprint-compiler/reference/objects.html) do it is a good idea. It removes the need for the name attribute and would also remove the old `method: name = Widget` syntax.

```rust
gtk::Box my_box {
    gtk::Label my_label {
        // ...
    }
}
```

### Placing methods

Relm4's syntax is similar to Rust (struct initialization), JSON and blueprint which makes it easy for beginners to understand the macro and also makes it very consistent. Therefore, I don't like the old C-style of doing things in the opposite order and rather like to have the type after the method name. I think that the idea with the argument specification is really nice though.

```rust
gtk::Box my_box {
    append(&#) = gtk::Label my_label {
        // ...
    },
    // It probably makes sense to add a shorter syntax and 
    // automatically add a reference because that covers at least 90% of all use-cases
    append = gtk::Label my_label {
        // ...
    }
}
```

### Builder pattern syntax

I really like the idea of making it easier to use builder patters, but I'm still not convinced by the approach declarative uses. While waiting for my train, I thought of another interesting approach: We could just allow the user to return a builder type (or any other type) by using `gtk::Box::builder()` instead of just `gtk::Box`. And then, the user could convert this builder into the final type by using something like `#build()` or `#other_method_name()`. Maybe this isn't the perfect solution yet, but I like its explicitness.

```rust
relm4::view! {
    adw::Flap::builder() my_flap {
        flap_position: gtk::PackType::End,
        fold_policy: adw::FlapFoldPolicy::Never,
        
        gtk::Box the_flap {
            // content
        },
        gtk::Box the_content {
            // content
        },
        // FlapBuider -> Flap
        #build(),
        set_flap_position: //...
    }
}
```

### Attributes

I don't have a lot of time, so I'll just write down my thoughts quickly. I think `bind` does a lot of things better than `watch` and `track`, so ideally Relm4 should use something similar. Also, maybe we can have something that makes it possible to apply a binding condition (or any attribute / label) to multiple properties without parenthesis. Not sure if that makes sense though, I will continue thinking about this tomorrow.

```rust
gtk::Box {
    @update
    set_margin_all: model.margin,
    
    gtk::Label {
        @update if let Some(text) = &model.text
        set_label: text,
    },
    gtk::Label {
        // Apply update to both properties
        @update if model.update_label
        > set_label: model.label,
        > set_margin_all: model.label_margin,
    }
}
```

**Reply by ejaa3** [reply]
Well, here are some reasons why `declarative` is so. It may be a lot of detail to process in such a short time, but perhaps there is less to discuss now.

## View syntax

> When looking at the rust lang reference, it becomes clear that... should have only few additional parameters

I don't see it clearly. It just says “_The first TokenStream is the delimited token tree following the attribute's name, not including the outer delimiters_”, and that is exactly what the `#[declarative::view]` does.

> declarative even breaks syntax highlighting in some editors because nobody even considered the possibility to (ab)use macros like this

The only way for a macro to “break” syntax highlighting is because it has hygiene issues. The hygiene information also allows for proper semantic highlighting, and in fact it is Relm4 who is breaking the syntax unlike `declarative`. Here a test image:

![hygiene](https://github.com/Relm4/Relm4/assets/108014240/ee00601d-78f1-421f-b6a0-1b36f6a22149)

You may want to open the image in a new tab to see it larger (100 % zoom to avoid blur).

I'm using VS Code and I have it configured to highlight methods in blue and modules in pink, but in Relm4 weird things happen (I point it with red arrows). The first problem is the arrow pointing to `gtk`, where `gtk` should be pink. When I mouse over there, I get a tooltip about `Default::default` instead of the `gtk` module and that's wrong. The second arrow is the property methods (like `set_titlebar`), which are not colored blue and have no tooltip. The third problem is signals (like `connect_clicked`) that are not colored blue, and I get a tooltip about `clippy::redundant_clone` (nothing to see).

Of course, there is also non-semantic syntax highlighting, which is usually based on regular expressions. And an incomplete regular expression may fail to highlight a `#[declarative::view]`, as in the [sample code in the crates.io README](https://crates.io/crates/declarative). However the [sample code on GitHub](https://github.com/ejaa3/declarative) looks good.

VS Code has both ways of highlighting: semantic and not, and I think the non-semantic uses regular expressions. And I found that non-semantics is poor when it comes to highlighting attributes like `declarative::view`, and so I use a theme that only highlights attributes semantically, more or less.

The point is that `declarative` should only be hygienic so that it is possible for the language server to provide semantics. Whether editors or their visual themes manage to perfectly highlight the syntax, whether based on semantics or regular expressions, is the problem of the editor, not `declarative`.

> I don't know a single crate that does things differently except for the current version of declarative

Maybe no one knew that attribute macros were pretty powerful. Now with `declarative` we have a pretty clear example.

## Wrapping in a trait

I'm not sure. `declarative` only tries to transform imperative code into declarative, and that is possible without traits. It should be enough with:

~~~ rust
impl AnyThing for AnyType {
    #[view {
        // widgets
    }]
    
    fn function() {
        expand_view_here! { }
    }
}
~~~

The same with a `mod` (happens to me):

~~~ rust
mod module {
    #[view {
        // widgets
    }]
    
    fn function() {
        expand_view_here! { }
    }
}
~~~

I can also apply `#[view]` as attribute to `impl` or `mod`, but while closer to `expand_view_here!` better. If you suggested traits as a workaround for syntax highlighting, I would insist that that is a code editor problem. It is still possible to use `block!`.

## Placing methods

You might find the `append(&#) = Widget { }` syntax consistent with the usual method calls, but…

~~~ rust
gtk::Box box_name {
    // this seems “consistent” with the `gtk::Box` above:
    adw::Bin bin_name #append(&#) {
        // content
    }
}
~~~

In addition, sometimes it will happen that the interpolation is long, [like here](https://github.com/ejaa3/declarative/blob/main/examples/h_no_channels.rs#L105). And I see better read this:

~~~ rust
gtk::Grid {
    Child::new("Name", this.clone()) my_child {
        #attach(&#.data.get().unwrap().0, 1, 2, 3, 4)
        // properties
    }
}
~~~

Than this:

~~~ rust
gtk::Grid {
    attach(&#.data.get().unwrap().0, 1, 2, 3, 4) = Child::new("Name", this.clone()) my_child {
        // properties
    }
    
    attach( // or maybe so
        &#.data.get().unwrap().0, 1, 2, 3, 4
    ) = Child::new("Name", this.clone()) my_child {
        // properties
    }
    
    attach(&#.data.get().unwrap().0, 1, 2, 3, 4) // or so
    = Child::new("Name", this.clone()) my_child {
        // properties
    }
}
~~~

In `match` it's easy to have the type next to the arrow:

~~~ rust
gtk::Grid {
    match variant {
        Variant::Switch => Child::new("Name", this.clone()) {
            #attach(&#.data.get().unwrap().0, 1, 2, 3, 4)
            // properties
        }
    }
}
~~~

You can see that the `#attach` inside looks like the [widget layouts in Blueprint](https://jwestman.pages.gitlab.gnome.org/blueprint-compiler/reference/extensions.html#widget-layouts). With the Relm4 syntax it would be:

~~~ rust
gtk::Grid {
    match variant {
        Variant::Switch => attach(
            &#.data.get().unwrap().0, 1, 2, 3, 4
        ) = Child::new("Name", this.clone()) {
            // properties
        }
    }
}
~~~

Also `#this_form(&#)` handles returning at any position within scope if followed by `'back`. With the Relm4 syntax I could only use `'back` at the beginning (in scope) or at the end (out of scope), because `'back` is supposed to edit what is returned by the function or method that was called “back”, unless one more keyword is dedicated just to edit the return of a composition.

## Out of curiosity...

I use the `#` sign at the beginning not only to differentiate an interpolation, but because the same symbol is used for attributes. So I can't apply attributes to an interpolation like this:

~~~ rust
gtk::Box my_box {
    #[first]
    adw::Bin my_bin {
        #[second]
        #append(&#) // syn will send an error on `append`
        'back returned { set_name: "name" } // suppose it returns
        
        @gtk::Box::append(&#, &returned)
    }
}
~~~

Basically because it would expand like:

~~~ rust
let my_box = gtk::Box::default();

#[first]
let my_bin = adw::Box::default();

#[first] // outermost attributes apply internally
#[second]
let returned = my_box.append(&my_bin);

#[first]
#[second]
returned.set_name("name");

#[first]
// here #[second] does not apply because the @extension is not inside 'back
// however it did apply to `returned`, and that is a severe error
@gtk::Box::append(&my_box, &returned);
~~~

If `#[second]` were a `#[cfg]`, `returned` would be optional, but in case of missing it, the app would not compile because there would be an error in `@extension(#)` (there is no `return` variable). So it is simply not possible to do that, and should rather be written:

~~~ rust
gtk::Box my_box {
    #[first]
    #[second]
    adw::Bin my_bin {
        // syn does not allow attributes for something starting with #
        #append(&#) 'back returned { set_name: "name" }
        @gtk::Box::append(&#, &returned)
    }
}
~~~

And now the `@extension(#)` would be just as optional as `returned`.

That is, I take advantage of the fact that syn does not allow attributes for something that begins with `#`, for precisely something that should not accept attributes, and the `#interpolation(#)` syntax seems ideal.

## Trimming syntax by 4 characters: `(&#)`

> It probably makes sense to add a shorter syntax and automatically add a reference because that covers at least 90% of all use-cases

I think there is no way. Assuming your syntax, `declarative` already has uses for `:`, `=`, `:=`, `;`, `;;` and `=>` (`->` is free). For example:

~~~ rust
Struct {
    field = "value"
    method: "arg1", "arg2"
    fn_field := "arg1", "arg2"
    
    field => { /* edit as to `Struct` */ }
    method_no_args;
    fn_field_no_args;;
}
~~~

To add items (as widgets), the first three cases would have to be extended like this:

~~~ rust
Struct {
    field[#] = Item { /* edit */}
    method[#]: Item { /* edit */}
    closure_field[#] := Item { /* edit */}
}
~~~

The difference would be the brackets. If I remove the brackets, it just goes back to the beginning and it wouldn't work. Of course for Relm4 there are plenty of punctuations because it only calls methods, but `declarative` also takes into account fields and functional fields.

Currently in `declarative` it is done like this:

~~~ rust
Struct {
    Item #field {#} { /* edit */}
    Item #method(#) { /* edit */}
    Item #closure_field[#] { /* edit */}
}
~~~

The difference is basically the use of braces, parens and brackets. So if I try to do `#something` (without parens) to fall back to `#something(&#)`, what will actually happen is that the macro will start looking for a `#` inside the curly braces that edit the widget thinking I'm assigning a field (unless there is a `!` before the brace), and it seems stingy to save a simple `(&#)` (better with it, more explicit).

## Builder pattern syntax

> We could just allow the user to return a builder type (or any other type) by using gtk::Box::builder() instead of just gtk::Box

Already `declarative` allows that. With the `gtk-rs` feature, if I just write the type it will autocomplete with `::builder()`, but if I write a full expression (like `gtk::Box::builder()`) it won't autocomplete.

> And then, the user could convert this builder into the final type by using something like #build() or #other_method_name().

I would say that `declarative` also allows that with the `gtk-rs` feature. For example:

~~~ rust
gtk::Box !{ // autocomplete
    spacing: 10 #: // autocomplete
    set_spacing: 20
}

// the above is the same as:
gtk::Box::builder() !{ // does not autocomplete
    spacing: 10
    build; #; // does not autocomplete
    set_spacing: 20
}

// both expand recursively to:
let _box = gtk::Box::builder().spacing(10).build();
_box.set_spacing(20);
~~~

In your example you forgot to put `!` before the brace. That symbol makes the methods be().chain().so(). However, if the mode is not finished manually with `#:` or `#;`, it will be autocompleted with `.build()`.

Now if you prefer to always be explicit (`builder()` and `build()`), you can bypass the `gtk-rs` feature and enable only `builder-mode` and implement the `builder_mode!` macro that is called in the background differently (in fact you wouldn't even need to activate `builder-mode` to avoid having to implement the macro, if you just want to always be explicit). You can also do as in Relm4:

~~~ rust
gtk::Box::builder().spacing(10).build() {
    set_spacing: 20
    // no composition with builder methods here
}
~~~

However I have a code like this:

![non-explicit](https://github.com/Relm4/Relm4/assets/108014240/47c72e98-1801-4754-8537-21148c985798)

If I were to write `builder()` and `build; #;` all the time, it would look like this (note the `Type::builder()` instead of just types, and the bunch of `build; #;` below):

![explicit](https://github.com/Relm4/Relm4/assets/108014240/374d90de-ae39-48d7-8fb0-9604bcb95fff)

## `'bind` braces

> maybe we can have something that makes it possible to apply a binding condition (or any attribute / label) to multiple properties without parenthesis

If for `match` I will always use braces (`'bind` supports `match`):

~~~ rust
'bind match variant {
    First(value) => set_value: value
    Second(value) => {
        set_value: value
        set_tooltip: "second"
    }
}
~~~

For `if` too:

~~~ rust
'bind if let First(value) = variant {
    set_value: value
} else if let Second(value) = variant {
    set_value: value
    set_tooltip: "second"
} else { /* ... */ }
~~~

It is consistent with `if` and `match` without `'bind` (the “static” or “non-dynamic” case).

**Comment by AaronErhardt** [comment]
### Wrapping in a trait

At least to me it looks pretty clear that *attribute* macros are supposed to extend built-in *attributes*. And Rust's syntax for attributes is in fact very limited as the [reference](https://doc.rust-lang.org/reference/attributes.html) and the [syn crate](https://docs.rs/syn/latest/syn/struct.Attribute.html) show. Just because we get a `TokenStream` it doesn't mean that we also should use it to implement arbitrary syntaxes. 

Therefore and also because there are no big downsides, I'd prefer a trait implementation or maybe we could actually use modules as well. 

### Placing methods

If I understand you correctly, there is no good reason against putting the method first. The idea of allowing the user to specify the interpolation within the bracket is pretty good as you demonstrated.

### Builder pattern syntax

> In your example you forgot to put ! before the brace. 

No, the `!` is not necessary. If somebody uses `#;` or any other distinct syntax, we already know that everything before is part of the builder pattern. Only the auto-completiong of `#;` would not be possible in this case.

**Reply by ejaa3** [reply]
## Wrapping in a trait

> Attribute macros are supposed to extend built-in attributes

If by that you tried to say that attribute macros are not supposed to have that much content, I think that's just your assumption.

> And Rust's syntax for attributes is in fact very limited as the reference...

I just read in the reference: “_The attribute consists of a path to the attribute, followed by an optional delimited token tree whose interpretation is defined by the attribute_”

In `#[view { /* widgets */ }]`, the “_path_” is `view` and the “_optional delimited token tree..._” is `{ /* widgets */ }`.

The crate syn fits that. That is, the `List` variant of the [`enum Meta`] takes a [`MetaList`], whose fields are precisely the path, and for the delimited token tree” it dedicates a delimiter field and an arbitrary `TokenStream`.

[`enum Meta`]: https://docs.rs/syn/latest/syn/enum.Meta.html#variant.List
[`MetaList`]: https://docs.rs/syn/latest/syn/struct.MetaList.html

However I could think of implementing something like `#[declarative::scope]`, which can be attributed to a `mod module`, an `impl Type` or an `impl Trait for Type` to use a single `view!`, with the only advantage of being able to write the view after the function that invokes `expand_view_here!` (no need to define some trait in `declarative`).

## Placing methods

> If I understand you correctly, there is no good reason against putting the method first

So you misunderstood me. I tried to say that there is no good reason in favor of putting the method before the type, except that it looks like Blueprint and JSON. And implementing both ways is just unnecessary maintenance.

## Builder pattern syntax

> If somebody uses `#;`..., we already know that everything before is part of the builder pattern

You know that because your eyes can see everything in braces as if nothing, but the macro doesn't know what's coming unless it unnecessarily constantly scans a lot of code ahead for a `#;`, and then discard it to parse it again, this time knowing that a `#;` is coming.

But there are even times where I don't even need to put a `#:` at the end like in a [previous image] I sent (lots of `!` but only one `#:`). Good luck implementing builder mode in Relm4 without `!`.

[previous image]: https://user-images.githubusercontent.com/108014240/241118240-47c72e98-1801-4754-8537-21148c985798.png

Now, from the start I didn't like the `#:`, `#;`, `@:` and `@;` markers, but I did it that way because it was quick to do and think about. Now I'm thinking that I can take advantage of the fact that builder methods can't accept attributes (`declarative` throws an error about it) and overload the use of `#`, perhaps like in the example you gave (`#build()`). It would look more or less like this:

~~~ rust
// the below cases expand to
adw::Bin::builder().margin_start(4).margin_end(4).build()

adw::Bin !{ // or adw::Bin::builder() !{
	margin_start: 4 // chained
	
	// first case: autocompletion
	#margin_end: 4
	
	// second case: no autocompletion
	margin_end: 4
	#:build;
	
	// third case: autocompletion but expands at the beginning
	#!margin_end: 4
	
	// fourth case: no autocompletion but expands at the beginning
	margin_end: 4
	##build;
}
~~~

The previous cases but with a child:

~~~ rust
// the below cases expand to:
adw::Bin::builder().margin_start(4).child(&_gtk_box).build()

adw::Bin !{
	margin_start: 4
	
	#gtk::Box #child(&#) { } // first case
	
	gtk::Box #child(&#) { }
	#:build; // second case
	
	#!gtk::Box #child(&#) { } // third case
	
	gtk::Box #child(&#) { } // fourth case
	##build;
}
~~~

The only difference between the first two cases and the last two is the position where they are expanded (sometimes it is necessary to place a builder at the beginning). The cases are ordered in descending order of usage probability, which is why I think those symbols are fine.
